tlog-cosignature: Allow additional statements

rgdd opened the pull request:
This PR puts a maintainer's touch on https://github.com/C2SP/C2SP/pull/140 by targeting continued use of cosignature/v1 rather than going up to cosignature/v2. This does not mean we did not like the suggested cosignature/v2 changes (witness name is tied to the cosignature and can be used as a signing context), but rather that we want to be sure there aren't other things we also want in a cosignature/v2. One way to view this is: while it's a painpoint to manage multiple keys for different cosigning contexts, it's also a painpoint to manage multiple different cosignature versions. So we want to avoid creating a cosignature/v2 prematurely and start by making a backwards compatible update to the text in cosignature/v1.

Blocking merge:

* Input from @davidben, @lukevalenta: are you happy with this update / do you agree it is sufficient?
* Indications that all three maintainers think this is the right way to move forward, otherwise $discuss.

Nice to have before merge:

* A v1.0.0 tag, so that this can become v1.1.0.

Finally -- many thanks for the great C2SP contributions davidben/lukevalenta!

(If anyone else has input that's of course also welcome. I already gathered input from a few Sigsum contributors that basically gave thumbs up when I described the changes that were brewing here.)

rgdd commented on tlog-cosignature.md:
```diff
@@ -71,34 +80,55 @@ over the message defined in the next section.
+
+A witness operator MAY make additional statements about a checkpoint.  These
+additional statements need to be communicated out of band to those defining
+trust policies based on using a unique witness name or a unique witness key.
```
> So I think that the name can be useful only when taken together with the key.

That's the intent, I'll see what I can do to clarify tomorrow. Thanks for the comment!

rgdd replied:
Updated to clearly specify that it is the tuple (witness name, public key, cosignature/v2 algorithm) that needs to be distinct for disambiguation, and that it's okay to use the same key with multiple different names (unlike in v1). 

Let me know what you think and if you consider your comment addressed!

eliasrudberg replied:
@rgdd thanks, I think it's much better now.

I think it's good to spell it out like you have done in one place, like this:

> tuples of (witness name, public key, supported cosignature version)

Then it's clear that what goes into a trust policy is information about the supported (required?) cosignature version.

In two other places the text says 

> tuples of (witness name, public key, cosignature/v2)

and that for me becomes confusing because it sounds like an actual consignature is part of the tuple, which could lead to some misunderstanding. So I think it would be better to write the whole "tuples of (witness name, public key, supported cosignature version)" in the other places also, or find some better way to express it.

What I mean is that when you write "(witness name, public key, cosignature/v2)" then a reader might take that literally and think that the tuple consists of:
- a witness name
- a public key
- a cosignature/v2

when in fact the tuple really consists of:
- a witness name
- a public key
- information about the required cosignature version

I think that I do understand what you mean, I just think there is a risk of misunderstanding there.

Another thought is that maybe instead of "supported cosignature version" it would be better to say "required cosignature version". My thinking is that someone writing a trust policy will write what is *required* for verification. But I'm not sure, I guess it depends on how you look at things.

mhutchinson commented on tlog-cosignature.md:
```diff
@@ -71,34 +80,57 @@ over the message defined in the next section.
+
+## Changes compared to cosignature/v1
+
+* The witness name is cryptographically tied to the resulting cosignature.
+* It is explicitly allowed to make additional statements about a checkpoint.
+  These additional statements need to be communicated out of band, identified by
+  the tuple (witness name, public key, cosignature/v2) in trust policies.  In
+  cosignature/v1, the public key cannot be reused with different witness names.
```
Can you add some justification for _why_ these changes are useful. The main purpose behind this seems to be to allow the same key to be used in different contexts (by setting different witness names). Why is embedding the name better than using different keys for different types of claim?

mhutchinson commented on tlog-cosignature.md:
```diff
@@ -95,10 +99,23 @@ log identified by the origin line has the specified root hash.
+A witness operator MAY make additional statements about a checkpoint.  These
+additional statements need to be communicated out of band to those defining
+trust policies based on tuples of (public key, supported cosignature version).
+The same tuple MUST NOT be reused across different cosigning contexts.
```
"different cosigning contexts" is potentially ambiguous; from our discussions I know that a cosigning context fixes a set of claims that can be relied on. But this sentence could be interpreted to mean something like "each key should only be used for cosigning a single log".

mhutchinson replied:
Pulled the trigger before reading the next sentence...

rgdd commented on tlog-cosignature.md:
```diff
@@ -95,10 +99,22 @@ log identified by the origin line has the specified root hash.
+A witness operator MAY make additional statements about a checkpoint.  These
+additional statements need to be communicated out of band to those defining
+trust policies based on tuples of (public key, supported cosignature version).
+The same tuple MUST NOT be reused across different cosigning contexts, where
```
How about that? Basically your suggestion but sticking to "statement" instead of "claim". Mostly because that's the term we've been using in the spec so far, and i think claim runs the risk of being preloaded with other thoughts due to e.g. the claimant model.

(And thanks -- very good suggestions!)

rgdd commented:
Squashed up the fixup commits and added a thanks in the commit message to @mhutchinson and @eliasrudberg for helping me tinker on the text in this PR.

davidben reviewed (approved):
Works for me!  Thanks! Yeah, Filippo mentioned he was leaning that way actually before I put together the PR. I only left it alone just to illustrate the v2 approach and figured you all can decide what you prefer. I'm happy with either.

Once this lands, I can go update the tlog-mirror PR and also change some of the terminology in Merkle Tree Certs.

In a vacuum, I think binding the witness name would have been good, but it's not critical, and it totally makes sense to not want to rev the signature construction for a change like that. (Plus, for MTCs, I expect we'd be reusing the concepts but defining our own signature construction anyway, so I especially don't have a horse in this race. Nothing against the C2SP construction. Just fallout from figuring out how to thread the C2SP/IETF needle because our signatures will intersect with the X.509 bits. Also we have this goofy [subtrees](https://davidben.github.io/merkle-tree-certs/draft-davidben-tls-merkle-tree-certs.html#name-subtrees) optimization.)

Regarding terminology, I noticed you also ended up with "witness" instead of "cosigner". That's fine by me! I interpreted "witness" to specifically mean the consistency-proof-following construction, but I was just guessing. Happy to just defer to you all on how to define these. :-)

Some misc thoughts on that, just to make sure we're on the same page:

I guess the thinking then is we'd say things like "a mirror is a witness that additionally promises to have mirrored the log", etc. Is that right?

I think another consequence is that a "witness" might not *necessarily* implement the protocol in tlog-witness, just any service that generates cosignatures meeting the guarantees. (It happens that tlog-mirror implements a modified witness protocol anyway, because this funky two-stage commit thing seemed valuable, but I think it would have been reasonable for one to not do that.)

One last thing is the relationship between cosignatures and log signatures. Now, in C2SP you all already separated them and I wouldn't propose retroactively merging them for C2SP. But for MTCs, I actually ended up merging the two.

The observation that log signature semantics are _also_ compatible with these new witness semantics. A witness signature isn't that it checked consistency proofs, but that every checkpoint it ever signs is consistent with each other. The log operator promises that too, just by virtue of knowing what it signed.

The merger then provided a nice hook (the additional statements framework) for the extra promise MTCs wanted to embed, which is "I have certified every entry in this checkpoint". It also means the relying party's verification process is very tidy and uniform ("look for sufficient cosignatures to satisfy you—you should look for enough to make you happy about authenticity and transparency but the core protocol is pretty agnostic to all this"). It also meant we didn't have to define subtree signatures twice, etc.

It also makes it very easy for log operators, as a robustness measure, do something like not actually expose the core log key, but instead operate a pair of (log, witness/mirror), with the witness/mirror key being the exposed log key. That way you defend a bit against accidental mishaps.

So, with that context, either in the MTC world which has to define a new signature construction anyway, or in a hypothetical alternate universe where tlog had merged the two... would "witness" still be a reasonable term for this even further generalized witness-or-mirror-or-log-owner signer?

rgdd commented:
> Works for me! Thanks! Yeah, Filippo mentioned he was leaning that way actually before I put together the PR. I only left it alone just to illustrate the v2 approach and figured you all can decide what you prefer. I'm happy with either.

Great -- thanks for the input!

> 
> Once this lands, I can go update the tlog-mirror PR and also change some of the terminology in Merkle Tree Certs.
> 
> In a vacuum, I think binding the witness name would have been good

100% agree, if/when there's a v2 I will advocate for this.

> , but it's not critical, and it totally makes sense to not want to rev the signature construction for a change like that. (Plus, for MTCs, I expect we'd be reusing the concepts but defining our own signature construction anyway, so I especially don't have a horse in this race. Nothing against the C2SP construction. Just fallout from figuring out how to thread the C2SP/IETF needle because our signatures will intersect with the X.509 bits. Also we have this goofy [subtrees](https://davidben.github.io/merkle-tree-certs/draft-davidben-tls-merkle-tree-certs.html#name-subtrees) optimization.)

I'll probably provide some comments somewhere else (Filippo provided hints on where it would be helpful), but what I can say right away is I did read the latest Merkle tree certificates draft and I liked the overall direction.

Pretty sure that the maintainers of c2sp.org's tlog specs are all interested in shaping any specs that have to be in IETF for $reasons. We've put quite a lot of effort in trying to converge the ecosystem rather than splitting it, and so far we've been very successful at that. I'm sure we'll be talking more to figure out what's the most helpful way to move things forward. 

> 
> Regarding terminology, I noticed you also ended up with "witness" instead of "cosigner". That's fine by me! I interpreted "witness" to specifically mean the consistency-proof-following construction, but I was just guessing. Happy to just defer to you all on how to define these. :-)

This is actually closer to my mental model. I.e., when I think of a witness, I think of a "cheap" append-only verifier that doesn't have to see the entire log (as per the tlog-witness spec). And if all goes well, produce a cosignature.

> 
> Some misc thoughts on that, just to make sure we're on the same page:
> 
> I guess the thinking then is we'd say things like "a mirror is a witness that additionally promises to have mirrored the log", etc. Is that right?

Yes. But the more I think about this I don't think it's helpful, see the bottom of this response.

> 
> I think another consequence is that a "witness" might not _necessarily_ implement the protocol in tlog-witness, just any service that generates cosignatures meeting the guarantees. (It happens that tlog-mirror implements a modified witness protocol anyway, because this funky two-stage commit thing seemed valuable, but I think it would have been reasonable for one to not do that.)
>

This is consistent with how I think of the specifications, i.e., tlog-witness should depend on tlog-cosignature but not vice versa.

> One last thing is the relationship between cosignatures and log signatures. Now, in C2SP you all already separated them and I wouldn't propose retroactively merging them for C2SP. But for MTCs, I actually ended up merging the two.
>

I don't see a problem with this, assuming proper namespaces so it's clear what's signed and who's signing (such as in the proposed cosignature/v2).

We did not consider this a few years ago because we optimized for backwards compatibility with existing things like Go's checksum database. I appreciate the observation though, nice one!

> The observation that log signature semantics are _also_ compatible with these new witness semantics. A witness signature isn't that it checked consistency proofs, but that every checkpoint it ever signs is consistent with each other. The log operator promises that too, just by virtue of knowing what it signed.
> 
> The merger then provided a nice hook (the additional statements framework) for the extra promise MTCs wanted to embed, which is "I have certified every entry in this checkpoint". It also means the relying party's verification process is very tidy and uniform ("look for sufficient cosignatures to satisfy you—you should look for enough to make you happy about authenticity and transparency but the core protocol is pretty agnostic to all this"). It also meant we didn't have to define subtree signatures twice, etc.
> 
> It also makes it very easy for log operators, as a robustness measure, do something like not actually expose the core log key, but instead operate a pair of (log, witness/mirror), with the witness/mirror key being the exposed log key. That way you defend a bit against accidental mishaps.
> 
> So, with that context, either in the MTC world which has to define a new signature construction anyway, or in a hypothetical alternate universe where tlog had merged the two... would "witness" still be a reasonable term for this even further generalized witness-or-mirror-or-log-owner signer?

Not sure what a unified format would be called, but it's probably not the most important part. What is important is to have clear terminology for the different parties that use the format, because trust policies will need to be specified based on understanding what the signatures from these different parties mean.

The key words I'd expect in a trust policy (and what meaning I'd associate them with) would be something like:

log - authoritatively moves the tree head forward, publishes entries; append-only for anything else to make sense

mirror - happy to cosign the most recent tree head it knows about and managed to replicate from entries; append-only is a requirement to make sense

witness - happy to cosign the most recent tree head it knows about if append only

I would find it a bit akward and clunky to say "a mirror is a witness that also replicates the log". I would rather help people build the mental model that a witness is a cheap append-only verifier whereas a mirror is a log replicator that (obviously) needs to verify append-only to function correctly. Both can use cosignature/v1.

With this in mind, I think you were onto something when doing s/witness/cosigner in a bunch of places and I missed it. I'll make one more commit with this change and let others weigh in on the change too. For me, it's a good idea because it's easier to overload "cosigner" and couple it with a distinct name, compared to overloading witness and trying to bolt it on to a different name like "mirror". And I do think it is helpful to make it easy to find distinct (and ideally easy) names.

Thanks again for all the helpful suggestions / input / comments / questions / etc!

davidben commented:
> Pretty sure that the maintainers of c2sp.org's tlog specs are all interested in shaping any specs that have to be in IETF for $reasons. We've put quite a lot of effort in trying to converge the ecosystem rather than splitting it, and so far we've been very successful at that. I'm sure we'll be talking more to figure out what's the most helpful way to move things forward.

Of course! Yeah, the MTC draft was just what seemed the best starting point to present to both communities. I'm sure there'll be a lot of hopping between IETF and C2SP to converge on what makes everyone happy. :-)

> With this in mind, I think you were onto something when doing s/witness/cosigner in a bunch of places and I missed it. I'll make one more commit with this change and let others weigh in on the change too. For me, it's a good idea because it's easier to overload "cosigner" and couple it with a distinct name, compared to overloading witness and trying to bolt it on to a different name like "mirror". And I do think it is helpful to make it easy to find distinct (and ideally easy) names.

Works for me. New version still LGTM.

lukevalenta commented on tlog-cosignature.md:
```diff
@@ -89,16 +93,26 @@ leading zeroes. This value MUST match the `timestamped_signature.timestamp`.
+
+A cosigner MAY make additional statements about a checkpoint.  These
+additional statements need to be communicated out of band to those defining
+trust policies based on tuples of (public key, supported cosignature version).
```
This should be the same tuple as configured on clients above, right?
```suggestion
trust policies based on tuples of (cosigner name, public key, supported cosignature version).
```

lukevalenta reviewed (approved):
LGTM! I'm happy with sticking with cosignature/v1 for the time being.

FiloSottile commented on tlog-cosignature.md:
```diff
@@ -95,10 +99,20 @@ log identified by the origin line has the specified root hash.
+include the cosignature/v1 semantics, and MAY include other statements that
+are non-conflicting.  Examples of non-conflicting statements include "I also
+mirrored the log up until the checkpoint size" and "I also reproduced the sparse
+Merkle tree root in the extension lines".  See [tlog-mirror][] for an example.
```
Agreed, I think it is implied by one possible reading of the rest of the text, but making it explicit is better than not.

davidben commented:
Uploaded https://github.com/C2SP/C2SP/pull/149 to pick up the change in https://github.com/C2SP/C2SP/pull/146/files#r2177948486
